File: src/slang/runtime/DeferredMatchbox.java
package slang.runtime;

import slang.parse.AbbreviatedAstPrinter;
import slang.parse.Expr;
import slang.parse.Stmt;

import java.util.ArrayList;
import java.util.List;

public class DeferredMatchbox extends Matchbox {
    public List<Value> deferredArgs;

    public DeferredMatchbox(List<Stmt.Match> matches, Environment environment, List<Value> deferredArgs) {
        super(matches, environment);
        this.deferredArgs = deferredArgs;
    }

    @Override
    public Value apply(Interpreter interpreter, List<Value> args) {
        List<Value> newArgs = new ArrayList<>(deferredArgs);
        newArgs.addAll(args);
        return super.apply(interpreter, newArgs);
    }

    @Override
    public String toString() {
        String repr = environment.collapsedString();
        repr += " deferring ";
        repr += deferredArgs;
        repr += " ";
        repr += new AbbreviatedAstPrinter().print(new Expr.MatchBlock(matches));
        return repr;
    }
}

File: src/slang/runtime/DeferredMatchbox.scala
package slang.runtime

import slang.parse.AbbreviatedAstPrinter
import slang.parse.Expr
import slang.parse.Stmt

case class DeferredMatchbox(matches: List[Stmt.Match], environment: Environment, deferredArgs: List[Value]) extends Value with Callable {
  val matchbox: Matchbox = Matchbox(matches, environment)

  def apply(interpreter: Interpreter, args: List[Value]): Value = {
    matchbox.apply(interpreter, deferredArgs ++ args)
  }

  override def toString: String = {
    var repr = environment.collapsedString
    repr += " deferring "
    repr += deferredArgs.mkString(" ")
    repr += " "
    repr += new AbbreviatedAstPrinter().print(Expr.Matchbox(matches))
    repr
  }

  override def getType: String = "Matchbox"

  override def asString: String = toString

  override def isElemental: Boolean = false
}